### ISSUE TITLE
use Java 8 Date and Time API

### ISSUE THREAD

**Issue by darkv** [wocommunity/wonder#755]
Now that Wonder 7 uses Java 8 the new Date and Time API could be used instead of Joda library.

**Comment by spelletier** [comment]
The problem I see here is this will trigger major code changes / import fixes for those of us using them without real benefit I can see.

If we do such a release, I think we should also consider adding some other changes that break current API.

**Comment by paulhoadley** [comment]
I've thought about this too, and have also wondered about the burden on existing code. We use the Joda types for attributes perhaps not _heavily_ but certainly quite a bit. I haven't looked at the Java 8 Date and Time API yet. Can anyone give a rough estimate of how compatible they are, and how painful a transition like this might be for legacy code?

**Comment by darkv** [comment]
Just played around with that idea and started a branch to start implementing this topic. Have a look at #756

**Comment by darkv** [comment]
WOTextField has a _formatter_ binding that takes a specific Format instance. By this you can control what time/date class should be used i.e. if your text field either produces NSTimestamp, Joda or the new java.time objects. Then you have the _dateformat_ binding which makes WOTextField create a formatter object based on the given pattern itself.

I think the support for the new date objects should be restricted to the _formatter_ binding. _dateformat_ should be left to produce NSTimestamps as it used to be. How many of us are really using the _dateformat_ binding? Using _formatter_ with Format instances gives you way more control on date conversions.

**Comment by paulhoadley** [comment]
> How many of us are really using the dateformat binding?

Just because you asked, I do use it on `ERXWOString`. I set `er.extensions.ERXSession.autoAdjustTimeZone=true` so that _displayed_ timestamps are adjusted from UTC to the timezone in the user's `ERXSession`. I don't disagree with your proposal (in that I never use a `WOTextField` to collect a timestamp anyway), and I presume it won't affect `ERXSession.autoAdjustTimeZone()` anyway because (I think) that only works for `NSTimestamp`s.

**Comment by darkv** [comment]
> in that I never use a WOTextField to collect a timestamp anyway

Now I am curious about how you make your users input timestamps :) separate fields for day, month and year and then constructing the timestamp in code?

**Comment by paulhoadley** [comment]
Put simply, I do _everything I possibly can to avoid having a user enter a full-fledged timestamp_. Most of the time, all I want is a user to be able to specify a (1-day resolution) _date_ (a task for which a timestamp is particularly ill-suited), and I use a Joda `LocalDate` for this. (We have a custom component that uses a JQuery-based calendar-style widget to collect this.) In some situations where we've also needed a time of day (say, a start or end time for real world event), we've combined a `LocalDate` with a `LocalTime`, and used the calendar widget along with a custom textfield/drop-down component showing times with 1-hour resolution, but also taking arbitrary text input.

If some in-app _event_ (where, say, you conceivably need to record an unambiguous instant in time) has a system-generated timestamp, I usually model this with `NSTimestamp`, and that's where I'm using `ERXSession.autoAdjustTimeZone()`—displaying these timestamps to the user in her own timezone.

**Comment by paulhoadley** [comment]
Hey @darkv, how's this work going? Did you get it to a point where we can think about merging it in? I see quite a bit of activity on #756.

**Comment by darkv** [comment]
It should be already quite functional though I did not do extensive tests yet. I was stuck in ERRest integration when I had to pause work on it. The problem I encountered is that ERRestTest fails, though not because of changes due to #756. Those tests must have been failing for a long long time already… They fail because the generated JSON strings are compared to static String objects which don't match the result anymore.

The generated JSON includes attributes in an arbitrary order which can—and most of the time will—differ from the constant. Additionally an _id_ attribute is always added which will break the comparison too.

I wanted to fix those tests first before considering merging the pull request but had no time yet to pursue that topic or decide how the test value assertion could be done alternatively.

**Comment by darkv** [comment]
Sooo, finally ERRestTest is running with no failing tests. @paulhoadley if you want to take a look at it please let me know. We probably can merge this soon.

**Comment by paulhoadley** [comment]
I'm just taking a look now. I'd like to use this in a new project. I've hit a wall trying to get the new prototypes to display in WOLips. I've checked out your branch, I've got `ERPrototypes` open in my workspace—should Entity Modeler just pick up these new prototypes automagically? It doesn't seem to be. This is a Maven-based project, and I'm targeting 7.0.0-SNAPSHOT, which should match. What else do I need to do?
